File: libs/statemanager/src/main/kotlin/mono/state/command/MouseCommandFactory.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.state.command

import mono.actionmanager.RetainableActionType
import mono.graphics.geo.MousePointer
import mono.shape.selection.SelectedShapeManager.ShapeFocusType.SELECT_MODE_HOVER
import mono.shape.shape.Line
import mono.shapebound.InteractionPoint
import mono.shapebound.LineInteractionPoint
import mono.shapebound.ScaleInteractionPoint
import mono.state.command.mouse.AddLineMouseCommand
import mono.state.command.mouse.AddTextMouseCommand
import mono.state.command.mouse.LineInteractionMouseCommand
import mono.state.command.mouse.MouseCommand
import mono.state.command.mouse.MoveShapeMouseCommand
import mono.state.command.mouse.ScaleShapeMouseCommand
import mono.state.command.mouse.SelectShapeMouseCommand

/**
 * A factory of [MouseCommand].
 */
internal object MouseCommandFactory {
    fun getCommand(
        commandEnvironment: CommandEnvironment,
        mousePointer: MousePointer,
        commandType: RetainableActionType
    ): MouseCommand? = when (mousePointer) {
        is MousePointer.Down -> detectMouseCommandWithMouseDown(
            commandEnvironment,
            mousePointer,
            commandType
        )

        is MousePointer.Click ->
            if (commandType == RetainableActionType.IDLE) SelectShapeMouseCommand else null

        is MousePointer.DoubleClick -> null
        is MousePointer.Move,
        is MousePointer.Drag,
        is MousePointer.Up,
        MousePointer.Idle -> null
    }

    private fun detectMouseCommandWithMouseDown(
        environment: CommandEnvironment,
        mousePointer: MousePointer.Down,
        commandType: RetainableActionType
    ): MouseCommand {
        val interactionCommand = detectInteractionCommand(environment, mousePointer)
        if (interactionCommand != null) {
            return interactionCommand
        }

        return when (commandType) {
            RetainableActionType.ADD_RECTANGLE -> AddTextMouseCommand(false)
            RetainableActionType.ADD_TEXT -> AddTextMouseCommand(true)
            RetainableActionType.ADD_LINE -> AddLineMouseCommand()
            RetainableActionType.IDLE -> SelectShapeMouseCommand
        }
    }

    private fun detectInteractionCommand(
        environment: CommandEnvironment,
        mousePointer: MousePointer.Down
    ): MouseCommand? {
        val focusingShape =
            environment.getFocusingShape()?.takeIf { it.focusType == SELECT_MODE_HOVER }

        if (!mousePointer.isWithShiftKey &&
            focusingShape != null &&
            focusingShape.shape !in environment.getSelectedShapes()
        ) {
            // If focusing shape is not selected and not in multiple selection mode (Shift key),
            // clear all selected shapes and select focusing shape.
            environment.clearSelectedShapes()
            environment.addSelectedShape(focusingShape.shape)
        }
        val selectedShapes = environment.getSelectedShapes()
        if (selectedShapes.isEmpty()) {
            return null
        }

        val sharpBoundInteractionCommand = createShapeBoundInteractionMouseCommandIfValid(
            environment,
            environment.getInteractionPoint(mousePointer.pointPx)
        )
        if (sharpBoundInteractionCommand != null) {
            return sharpBoundInteractionCommand
        }

        if (!mousePointer.isWithShiftKey &&
            environment.isPointInInteractionBounds(mousePointer.boardCoordinate)
        ) {
            val relatedShapes = selectedShapes
                .asSequence()
                .flatMap { environment.shapeManager.shapeConnector.getConnectors(it) }
                .mapNotNull { environment.shapeManager.getShape(it.lineId) }
            return MoveShapeMouseCommand(selectedShapes, relatedShapes)
        }
        return null
    }

    private fun createShapeBoundInteractionMouseCommandIfValid(
        commandEnvironment: CommandEnvironment,
        interactionPoint: InteractionPoint?
    ): MouseCommand? {
        if (interactionPoint == null) {
            return null
        }
        val shape =
            commandEnvironment.shapeManager.getShape(interactionPoint.shapeId) ?: return null
        return when (interactionPoint) {
            is ScaleInteractionPoint -> ScaleShapeMouseCommand(shape, interactionPoint)
            is LineInteractionPoint ->
                if (shape is Line) LineInteractionMouseCommand(shape, interactionPoint) else null

            null -> null
        }
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/command/mouse-command-factory.ts
import { RetainableActionType } from "$mono/action-manager/retainable-actions";
import { type FocusingShape, ShapeFocusType } from "$mono/shape/selected-shape-manager";
import { Line } from "$mono/shape/shape/line";
import type { CommandEnvironment } from "$mono/state-manager/command-environment";
import { AddLineMouseCommand } from "$mono/state-manager/command/mouse/add-line-mouse-command";
import { AddTextMouseCommand } from "$mono/state-manager/command/mouse/add-text-mouse-command";
import { LineInteractionMouseCommand } from "$mono/state-manager/command/mouse/line-interaction-mouse-command";
import { MouseCommand } from "$mono/state-manager/command/mouse/mouse-command";
import { MoveShapeMouseCommand } from "$mono/state-manager/command/mouse/move-shape-mouse-command";
import { ScaleShapeMouseCommand } from "$mono/state-manager/command/mouse/scale-shape-mouse-command";
import { SelectShapeMouseCommand } from "$mono/state-manager/command/mouse/select-shape-mouse-command";
import {
    InteractionPoint,
    LineInteractionPoint,
    ScaleInteractionPoint,
} from "$mono/shape-interaction-bound/interaction-point";
import { MousePointer, type MousePointerDown, MousePointerType } from "$mono/workspace/mouse/mouse-pointer";

/**
 * A factory of [MouseCommand].
 */
export class MouseCommandFactory {
    static getCommand(
        commandEnvironment: CommandEnvironment,
        mousePointer: MousePointer,
        commandType: RetainableActionType
    ): MouseCommand | null {
        switch (mousePointer.type) {
            case MousePointerType.DOWN:
                return this.detectMouseCommandWithMouseDown(commandEnvironment, mousePointer, commandType);
            case MousePointerType.CLICK:
                return commandType === RetainableActionType.IDLE ? SelectShapeMouseCommand : null;
            case MousePointerType.DOUBLE_CLICK:
            case MousePointerType.MOVE:
            case MousePointerType.DRAG:
            case MousePointerType.UP:
            case MousePointerType.IDLE:
                return null;
        }
    }

    private static detectMouseCommandWithMouseDown(
        environment: CommandEnvironment,
        mousePointer: MousePointerDown,
        commandType: RetainableActionType
    ): MouseCommand {
        const interactionCommand = this.detectInteractionCommand(environment, mousePointer);
        if (interactionCommand) {
            return interactionCommand;
        }

        switch (commandType) {
            case RetainableActionType.ADD_RECTANGLE:
                return new AddTextMouseCommand(false);
            case RetainableActionType.ADD_TEXT:
                return new AddTextMouseCommand(true);
            case RetainableActionType.ADD_LINE:
                return new AddLineMouseCommand();
            case RetainableActionType.IDLE:
                return SelectShapeMouseCommand;
        }
    }

    private static detectInteractionCommand(
        environment: CommandEnvironment,
        mousePointer: MousePointerDown
    ): MouseCommand | null {
        const focusingShape = this.getHoveringFocusingShape(environment);

        if (!mousePointer.isWithShiftKey && focusingShape && !environment.getSelectedShapes().has(focusingShape.shape)) {
            environment.clearSelectedShapes();
            environment.addSelectedShape(focusingShape.shape);
        }

        const selectedShapes = Array.from(environment.getSelectedShapes());
        if (selectedShapes.length === 0) {
            return null;
        }

        const sharpBoundInteractionCommand = this.createShapeBoundInteractionMouseCommandIfValid(
            environment,
            environment.getInteractionPoint(mousePointer.clientCoordinate)
        );
        if (sharpBoundInteractionCommand) {
            return sharpBoundInteractionCommand;
        }

        if (!mousePointer.isWithShiftKey && environment.isPointInInteractionBounds(mousePointer.boardCoordinate)) {
            const relatedShapes = Array.from(selectedShapes)
                .flatMap(shape => environment.shapeManager.shapeConnector.getConnectors(shape))
                .map(connector => environment.shapeManager.getShape(connector.lineId))
                .filter(shape => shape !== undefined);
            return new MoveShapeMouseCommand(selectedShapes, relatedShapes);
        }
        return null;
    }

    private static getHoveringFocusingShape(environment: CommandEnvironment): FocusingShape | null {
        const focusingShape = environment.getFocusingShape();
        if (focusingShape?.focusType !== ShapeFocusType.SELECT_MODE_HOVER) {
            return null;
        }
        return focusingShape;
    }

    private static createShapeBoundInteractionMouseCommandIfValid(
        commandEnvironment: CommandEnvironment,
        interactionPoint: InteractionPoint | null
    ): MouseCommand | null {
        if (!interactionPoint) {
            return null;
        }
        const shape = commandEnvironment.shapeManager.getShape(interactionPoint.shapeId);
        if (!shape) {
            return null;
        }
        if (interactionPoint instanceof ScaleInteractionPoint.Base) {
            return new ScaleShapeMouseCommand(shape, interactionPoint);
        }
        if (interactionPoint instanceof LineInteractionPoint.Base && shape instanceof Line) {
            return new LineInteractionMouseCommand(shape, interactionPoint);
        }
        return null;
    }
}
